src/twisted/web/client.py

```diff
@@ -955,6 +955,112 @@ def creatorForNetloc(self, hostname, port):
                           WebClientContextFactory.__name__)
 
 
+class _HostnameCacheEntry(object):
+    """
+    Entry in the L{HostnameCachingHTTPSPolicy} associating a connection
+    creator and a cacheId.
+
+    @ivar creator: See L{__init__}
+    @ivar cacheId: See L{__init__}
+    """
+
+    def __init__(self, creator, cacheId):
+        """
+        @param creator: The connection creator corresponding to the hostname to
+            which this object is mapped.
+        @type creator: L{client connection creator
+            <twisted.internet.interfaces.IOpenSSLClientConnectionCreator>}
+
+        @param cacheId: The id of this entry in the cache
+        @type cacheId: L{int}
+        """
+        self.creator = creator
+        self.cacheId = cacheId
+
+
+
+@implementer(IPolicyForHTTPS)
+class HostnameCachingHTTPSPolicy(object):
+    """
+    SSL connection creator for web clients that caches the last cacheSize
+    L{client connection creators <twisted.internet.interfaces.
+    IOpenSSLClientConnectionCreator>} for reuse in subsequent requests.
+
+    @ivar _trustRoot: See C{trustRoot} parameter of L{__init__}.
+
+    @ivar _cache: A cache associating hostnames to their
+        L{_HostnameCacheEntry}.
+    @type _cache: L{dict}
+
+    @ivar _cacheSize: See C{cacheSize} parameter of L{__init__}.
+    @ivar _nextCacheId: The id to use for the next cache modification,
+        either a hit or a new entry.
+    @type _nextCacheId: L{int}
+    """
+
+    def __init__(self, trustRoot=None, cacheSize=20):
+        """
+        @param trustRoot: See the C{trustRoot} parameter of
+            L{optionsForClientTLS}.
+        @type trustRoot: L{IOpenSSLTrustRoot}
+
+        @param cacheSize: The maximum size of the hostname cache.
+        @type cacheSize: L{int}
+        """
+        self._trustRoot = trustRoot
+        self._cache = {}
+        self._cacheSize = cacheSize
+        self._nextCacheId = 0
+
+
+    def _purgeCache(self):
+        """
+        Purge cache entries if cacheSize is exceeded.
+        """
+        idToRemove = self._nextCacheId - self._cacheSize
```

Comment:
There's a bug here :frowning_face:  Because [adding item to the cache also increments `_nextCacheId`](https://github.com/twisted/twisted/pull/793/files#diff-83c0fea4716a44bf877b0b63efb6006aR1060), the distance between the most recently accessed and its successor grows with each access.  After `_cacheSize` accesses, `idToRemove` is larger than the largest ID.

Suppose I create a cache with 3 items:
```python
cache = HostnameCachingHTTPSPolicy(trustRoot="ignored", cacheSize=3)
cache.creatorForNetloc(b"host1", port=80)
cache.creatorForNetloc(b"host2", port=80)
cache.creatorForNetloc(b"host3", port=80)
```
Then, I ask for `b"host1"` 3 times in a row:
```python
for _ in range(3):
    cache.creatorForNetloc(b"host1", port=80)
```
I get an exception when I attempt to get a creator for a new host:

```python
cache.creatorForNetloc(b'host4', port=80)
# Traceback (most recent call last):
# File "bug.py", line 24, in <module>
#    cache.creatorForNetloc(b'host4', port=80)
#  File "/home/mrw/src/python/twisted/src/twisted/web/client.py", line 1055, in creatorForNetloc
#    self._purgeCache()
#  File "/home/mrw/src/python/twisted/src/twisted/web/client.py", line 1028, in _purgeCache
#    del self._cache[keyToRemove]
# UnboundLocalError: local variable 'keyToRemove' referenced before assignment
```

That's because `_nextCacheId` is 6 and the cache size is 3, so `idToRemove` is 3, but the largest cache ID any entry has is 2.

A solution would be to sort the cache keys in increasing order, then remove all but the last `self._cacheSize` items:

```python
def _purgeCache(self):
    """
    Purge cache entries if cacheSize is exceeded.
    """
    items = self._cache.items()
    items.sort(key=lambda kv: kv[1].cacheId)
    del items[:-self._cacheSize]
    self._cache = dict(items)
```

However, I recommend [using `OrderedDict` instead](https://github.com/jlitzingerdev/twisted/pull/1).

Reply:
Thanks, that's a good find, sorry about the bug.  That use of OrderedDict is fantastic!  I honestly did not even consider it because I didn't realize popitem had those semantics.  I'm more familiar with deque and namedtuple from collections...seems there's always something new to learn in the library.